### ISSUE TITLE
propagate `pytree_dataclass` typechecking through vectorizing transformations

### ISSUE THREAD

**Issue by zhangir-azerbayev** [MatX-inc/seqax#2]
A rough sketch of the way a transformer is implemented in `train.py` might be 
```python
@pytree_dataclass
class Model:
  w_up: f32['layers d_model/d d_ff/t']
  # ...
  
@typechecked
def transformer_block(x: f32[b'batch/d seq d_model/t'], w_block: Any) -> f32[b'batch/d seq d_model/t']:
  # ...
  return x, ()
  
@typed_shard_map
def transformer(x: f32[b'batch/d seq d_model/t'], w_model: Model) -> f32[b'batch/d seq d_model/t']:
  y, _ = jax.lax.scan(transformer_block, x, w)
  return y
```
It would be helpful if there was some way of checking that `w_block` had the expected type, i.e the same as `Model` but with the leading `'layers'` axis removed from each field. One very hacky way of doing this is to use
```python
@pytree_dataclass
class Block:
  w_up: f32['d_model/d d_ff/t']
  # ...
  
@pytree_dataclass
class Model(Block):
  w_up: f32['layers d_model/d d_ff/t']
  # ...
  
@typechecked
def transformer_block(x: f32[b'batch/d seq d_model/t'], w_block: Block) -> f32[b'batch/d seq d_model/t']:
# ...
```
However, in more complex situations, this quickly leads to a large amount of boilerplate dataclasses. Ideally, there might be some kind of api that would allow e.g
```python
Model = expand_named_axes('layers', Block)
```
If the maintainers think this is a good idea, I'm happy to start working on a draft implementation.

**Comment by reinerp** [comment]
This is a great suggestion! I completely share your concerns about the current approach and its issues, and I think something in the direction you propose could be nice.

I think intuitive syntax (instead of `expand_named_axes('layers', Block)`) would be nice, so as to evoke people's existing intuitions from type systems in other programming languages. The intuition I'd hope to evoke is "`Model` is an array of `Block`". The most natural way I can see to do that is `Model = Block['layers']` or maybe `Model = Array['layers', Block]`.

**Comment by zhangir-azerbayev** [comment]
Your proposed syntax is much better, I especially like `Model = Array['layers', Block]`.
